refactor: refactoring pipelines

Farenheith opened the pull request:
WIP

I've created some reusable actions to be added at build and publish ones.

The publish action is not finished. I've generated it automatically and now I need to take a look into it to adapt for this project. I gave you write permission to this repository in case you want to change something directly, too.

Farenheith commented:
@Exopandora I took a look and had to do one small adjustment. Maybe there's something else to fix about the command to publish, but I think you may know it better than me

Exopandora commented on .github/workflows/gradle-build.yml:
```diff
@@ -0,0 +1,55 @@
+          distribution: 'temurin'
+          cache: gradle
+
+      - name: Make Gradle wrapper executable
+        run: chmod +x ./gradlew
+      
+      - name: Make Gradle wrapper executable
+        run: chmod +x ./gradlew
```
This should only be needed once 😅

Farenheith replied:
> This should only be needed once 😅

Unintentional Ctrl + c Ctrl + v 😅
 I'll fix it tonight and I'll read carefully your publishing process

Exopandora reviewed (commented):
How the publishing process works:
1. Update the contents of `changelog.txt` in the root directory of the project (currently excluded by .gitignore, but that could be changed). I prefer to not autogenerate the changelog.
2. Run `gradlew publishMods`. This will publish the mod to both curseforge and modrinth. In order for this to work, the gradle properties `curse_api_key` and `modrinth_api_key` need to be set (see publishMods block in [fabric|forge|neoforge]/build.gradle.kts). To make this work in CI, the call to `findProperty` will most likely need to be changed to `System.getEnv("...")`, so it will load environment variables instead. It is also possible to publish the mod only to a single platform by either using `gradlew publishModrinth` or `gradlew publishCurseforge`. It is also possible to only publish a single modloader version by running the tasks only for a specific submodule. For example `gradlew :fabric:publishMods` would only publish the fabric version of the mod to modrinth and curseforge. It is totally fine if the pipeline only supports publishing all loader versions to all distribution platforms.

Farenheith commented:
Take a look at the generated code:

```
gradlew ${{ inputs.publish-task }} \
            -P${{ inputs.param-prefix }}_api_key=$PLATFORM_TOKEN \
            -P${{ inputs.param-prefix }}FileType=release \
            -P${{ inputs.param-prefix }}Changelog="See GitHub releases for changelog: https://github.com/${{ github.repository }}/releases" \
            ${{ inputs.extra-params }}
```

It's in the publish-platform.yml, and publish-task will be publishCurseForge for CurseForge, and publishModrinth for modrinth, which complies with what you explained above. The action steps will require approval, so when they start to run, you'll need to accept each publication to happen; this can be changed.
Now, take a look at the second parameter `-P${{ inputs.param-prefix }}_api_key=$PLATFORM_TOKEN`, the parameter name will translate to curse_api_key for CurseForge and modrinth_api_key for ModRinth. Will it be enough to recognize the keys the way they are?

The only thing left is the second and third parameters, which may be unnecessary, but you'll need to keep changelog on repo for it to work properly, as you said.

Now, you still need to establish how the publishing routine will work. If you access the action, you can trigger it manually, like the build one:
![image](https://github.com/user-attachments/assets/d8a44787-bc7f-45c1-a0d7-4e7c6087f7f2)
You can choose between tags or branches to trigger the pipeline.
![image](https://github.com/user-attachments/assets/d8d7bdd3-8906-4270-b122-26ccd9ff9158)

The action is also prepared to trigger publishing when you create a new tag, which I think is the best way to go, as you'll have good control over time with it.

I updated the PR, changing what I explained here. If you want, you can try to run the command that the action will execute in the next publishing you do to see if you need to change anything so it can work:

For example:
```
 ./gradlew publishModrinth -Pmodrinth_api_key=YOUR_API_KEY_HERE
```

Exopandora commented:
> The action steps will require approval, so when they start to run, you'll need to accept each publication to happen; this can be changed.

This sounds good. I think the best workflow would be to create a new publishing pipeline whenever a new tag gets pushed. Then, each pipeline needs to be triggered manually to actually release the build.

> Now, take a look at the second parameter -P${{ inputs.param-prefix }}_api_key=$PLATFORM_TOKEN, the parameter name will translate to curse_api_key for CurseForge and modrinth_api_key for ModRinth.

Passing properties with `-P` seems to work as intended. You can test publishing by setting `dryRun = true` in the `publishMods` blocks of `[fabric/forge/neoforge]/build.gradle.kts` like so:
```diff
publishMods {
+	dryRun = true
	displayName = "$jarName-NeoForge-${libs.versions.minecraft.get()}-$modVersion"
	....
}
```
Instead of publishing the artifact, it will print out some debug information. See plugin reference [here](https://github.com/modmuss50/mod-publish-plugin).

> Will it be enough to recognize the keys the way they are?

I think it would be better to name them `CURSE_API_KEY` and `MODRINTH_API_KEY`. This way, it would also be possible to have a joined publishing task.

> The only thing left is the second and third parameters, which may be unnecessary,

If you want to get the release parameter working, you need to adjust the `build.gradle.kts` for each modloader module and set the `type` field accordingly. The `type` field can also be set platform independent (which is always the case for this project). Example on how it could work:
```diff
publishMods {
-	type = STABLE
+	type = when (findProperty("release_type")?.toString()) {
+		"release" -> STABLE
+		"beta" -> BETA
+		"alpha" -> ALPHA
+		else -> STABLE
+	}
}
```

> but you'll need to keep changelog on repo for it to work properly, as you said.

I'll add the changelog.txt to the repository.

Thanks for working on this!

Exopandora commented:
You could also use environment variables directly instead of passing the parameter with -P. For this, you need to replace `findProperty("curse_api_key")` and `findProperty("modrinth_api_key")` with `System.getenv("CURSE_API_KEY")` and `System.getenv("MODRINTH_API_KEY")` respectively (casing change is only by convention).

Farenheith commented:
@Exopandora, what do you think is the best? If I change to use environment variables, this can affect the way you publish locally. You may need to fall back to this method if anything goes wrong with the integration. The -P would work with the way the code is now, right?

Exopandora commented:
> If I change to use environment variables, this can affect the way you publish locally.

I can adjust my local method, though I don't think I'll be using the local method again once this PR is merged.

> The way it is, you'll need to clear and update the changelog in place in your repository.

Yep. I'm totally fine with that. If I later decide to keep the changelog for older versions, I could do that by parsing the relevant bits of the changelog file, which can be done in the build script.

Farenheith commented:
@Exopandora ok, I made the changes and set the publishing action to run when opening a PR just to make it run and check if everything is ok before building. I'll remove this trigger after the test is done

Farenheith commented:
@Exopandora I've run the publish pipeline and it worked as expected. Of course, it didn't publish anything as the secrets are not set yet in your repository, but you can see in the image below how it went:

![image](https://github.com/user-attachments/assets/d8c40263-143b-4fbc-bc07-ea62d070e914)
Link [here](https://github.com/Exopandora/ShoulderSurfing/actions/runs/14744988619)

See the step "Matrix approval"? I've removed it from the flow for now, which means that, when you create a tag or trigger this pipeline, it'll automatically publish to CurseForge and Modrinth without any confirmation. If you want further approval, you'll need to create some environments in the settings of your repository requiring manual approval, and then update the pipeline. As I don't have access to it, I've removed it for now to simplify this PR.

I think it's ready to go. Please review the PR to make sure I haven't messed up anything. After approving, you'll need to add the secrets in your repository (CURSE_API_KEY and MODRINTH_API_KEY) and test the pipeline. Let me know when you do it so I can check if everything works as expected!

Exopandora commented on .github/workflows/publish-platform.yml:
```diff
@@ -0,0 +1,74 @@
+        required: true
+        type: string
+        description: "Gradle task to run for publishing"
+      param-prefix:
+        required: true
+        type: string
+        description: "Parameter prefix for the platform (curse, modrinth)"
```
If I'm not mistaken, this input can be removed.

Farenheith replied:
I'll fix it as soon I'm off work

Exopandora commented on .github/workflows/setup-java.yml:
```diff
@@ -0,0 +1,61 @@
+          UNIQUE_VERSIONS=$(echo "$ALL_VERSIONS" | tr ',' '\n' | sort | uniq | tr '\n' ',' | sed 's/,$//')
+          
+          echo "mc-versions=$UNIQUE_VERSIONS" >> $GITHUB_OUTPUT
+          echo "Mod version: $MOD_VERSION"
+          echo "Minecraft versions: $UNIQUE_VERSIONS"
```
It could be beneficial to append these to the environement aswell. For example, they could be used to generate the jar file name in the build task, so we don't have to rely on wildcard. But nothing critical.

Farenheith replied:
Maybe this is better to get from gradle.properties, don't you think?

Exopandora reviewed (commented):
> If you want further approval, you'll need to create some environments in the settings of your repository requiring manual approval, and then update the pipeline. As I don't have access to it, I've removed it for now to simplify this PR.

How exactly would I do that? And how can I protect the publishing trigger? For example, if someone creates a PR and changes the pipeline definition to trigger "on PR", it would publish a new version of the mod, with the contents of the PR.

Exopandora commented:
Also, it might be worthwhile to use the [gradle build cache](https://docs.gradle.org/current/userguide/build_cache.html) across multiple workflows.

Farenheith commented:
I see you've clarified your doubts about PR security already. I'll be honest about that: My team where I work uses environment approval control and, as we're all contributors in the repository, we can just change the pipeline and force a deploy. We don't do that because we like to be employed lol.

The scenario of open-sourced contribution brings some concerns I've never thought about. I have some NodeJS libraries with automated publishing too, and although I've never had any problems, I've also never checked GitHub security in this matter. I'll read the documentation to understand how it works now lol, but I'm glad you already checked it.

Farenheith commented:
@Exopandora I had some problems trying to enable cache in the first run. I need to take a time to understand it better. You can see the try [here](https://github.com/Exopandora/ShoulderSurfing/actions/runs/14762913825/job/41447642036?pr=298). Can it enter later a next improvement? That way we can assure the publishing will work the way it is first, so we have a trusted baseline.

Exopandora commented:
> Can it enter later as a next improvement? That way, we can assure the publishing will work the way it is first, so we have a trusted baseline.

Of course! Sounds like a very good idea.

Regarding the issues you were facing. I suspect that they come from enabling the **configuration** cache, which does not seem to be supported by all gradle plugins used in the project. You can leave that disabled, as it will not make much of a difference anyway, I think.

Thanks for the hard work! I will merge this PR now and backport the current changes on master to some other versions in the next days. When that is done, I'll let you know when I try out the publishing.

Farenheith commented:
Maybe it's a good idea to test it on matter before back porting, because if some error occurs, we can avoid the reworking of back porting again

Exopandora commented:
> Maybe it's a good idea to test it on master before back porting.

Makes sense. I'll try to trigger a release via tag on master later today.

Exopandora commented:
Publishing worked! Thank you! It would be really great to get the cache working, as the publishing step fully re-setups the project for each platform. Ideally, caches would be per branch.
